table/pit-cs.go

```diff
@@ -8,56 +8,86 @@
-
-	nPitEntries int // Number of PIT entries in tree
-	pitTokenMap map[uint32]*PitEntry
-
-	nCsEntries    int // Number of CS entries in tree
-	csReplacement CsReplacementPolicy
-	csMap         map[uint64]*CsEntry
-}
-
-// PitCsNode represents an entry in a PIT-CS tree.
-type PitCsNode struct {
-	component ndn.NameComponent
-	depth     int
-
-	parent   *PitCsNode
-	children []*PitCsNode
-
-	pitEntries []*PitEntry
-
-	csEntry *CsEntry
-}
-
-// PitEntry is an entry in a thread's PIT.
-type PitEntry struct {
-	node  *PitCsNode
-	pitCs *PitCs
-
-	Name           *ndn.Name
-	CanBePrefix    bool
-	MustBeFresh    bool
-	ForwardingHint *ndn.Name                // Interests must match in terms of Forwarding Hint to be aggregated in PIT.
-	InRecords      map[uint64]*PitInRecord  // Key is face ID
-	OutRecords     map[uint64]*PitOutRecord // Key is face ID
-	ExpirationTime time.Time
-	Satisfied      bool
-
-	Token uint32
+// PitCsTable dictates what functionality a Pit-Cs table should implement
+type PitCsTable interface {
+	InsertInterest(interest *ndn.Interest, hint *ndn.Name, inFace uint64) (PitEntry, bool)
+	RemoveInterest(pitEntry PitEntry) bool
+	// RemoveInterestByName(name *ndn.Name)
+	FindInterestExactMatch(interest *ndn.Interest) PitEntry
+	// FindInterestExactMatchByName(name *ndn.Name) PitEntry
+	FindInterestPrefixMatch(interest *ndn.Interest, token uint32) []PitEntry
+	FindInterestPrefixMatchByData(data *ndn.Data, token *uint32) []PitEntry
+	// FindInterestPrefixMatchByName(me *ndn.Name) []PitEntry
+
+	PitSize() int
+
+	// InsertOutRecord(interest *ndn.Interest, face uint64) *PitOutRecord
+	// GetOutRecords(interest *ndn.Interest) []*PitOutRecord
+	// GetOutRecordsByName(name *ndn.Name) []*PitOutRecord
+
+	InsertData(data *ndn.Data)
+	FindDataExactMatch(interest *ndn.Interest) CsEntry
+	FindDataPrefixMatch(interest *ndn.Interest) CsEntry
+	CsSize() int
+	IsCsAdmitting() bool
+	IsCsServing() bool
+
+	eraseCsDataFromReplacementStrategy(index uint64)
+
+	ExpiringPitEntries() chan PitEntry
+}
+
+// BasePitCsTable contains properties common to all PIT-CS tables
+type BasePitCsTable struct {
+	expiringPitEntries chan PitEntry
+}
+
+// PitEntry dictates what entries in a PIT-CS table should implement
+type PitEntry interface {
+	PitCs() PitCsTable
+	Name() *ndn.Name // external users can access this
+	CanBePrefix() bool
+	MustBeFresh() bool
+	ForwardingHint() *ndn.Name
+	// Interests must match in terms of Forwarding Hint to be
+	// aggregated in PIT.
+	InRecords() map[uint64]*PitInRecord   // Key is face ID
+	OutRecords() map[uint64]*PitOutRecord // Key is face ID
+	ExpirationTime() time.Time
+	SetExpirationTime(t time.Time)
+	Satisfied() bool
+	SetSatisfied(isSatisfied bool)
+
+	Token() uint32
+
+	InsertInRecord(interest *ndn.Interest, face uint64, incomingPitToken []byte) (*PitInRecord, bool)
+	InsertOutRecord(interest *ndn.Interest, face uint64) *PitOutRecord
+
+	GetOutRecords() []*PitOutRecord
+	ClearOutRecords()
+	ClearInRecords()
```

Comment:
Added `ClearInRecords()` and `ClearOutRecords()` to the `PitEntry` interface, they weren't in the doc originally.